### ISSUE TITLE
Number input model and $valid function not working when number incremented using stepUp() and stepDown()

### ISSUE THREAD

**Issue by Om1986** [angular/angular.js#15146]
I am trying to validate my angular form with a number input. I have styled my number input ( which i have not included in my demo) so that i use stepUp() and stepDown() to increment and decrement values.

Here's my demo : http://plnkr.co/edit/MTYvDP3XlID7IiJK0qIl?p=preview

form.numbah.$valid and the number's ng-model don't change when i used stepUp()/stepDown() to increment/decrement values. 

Alternatively, when you increment/decrement $scope.numbah from controller , the validation works fine. 

Ideally since the number input's values are changed , $valid and ng-model should also change in both the cases. But it is not happening. Can you please look into this issue?

**Comment by Narretz** [comment]
StepUp and stepDown don't seem to trigger the change event, so Angular has no idea that the input value changed. You should be able to propagate the change by calling $setViewValue with the current input value after you have called the methods.

**Comment by kathir007** [comment]
maghilchi

**Comment by gkalpak** [comment]
Everything seems to work as expected here. Should this be closed, @Narretz?

**Comment by Om1986** [comment]
@Narretz stepUp() and stepDown() are native javascript functions. So i don't understand , why Angular should not support these events ?

**Comment by gkalpak** [comment]
As you said, they are not events, they are methods. `ngModel` detects changes in the element's value by listening to the `input` or `change` event (that are fired for example each time the value of the element changes as a result of the user's interaction with it).

Calling the `stepUp()`/`stepDown()` methods does not seem to produce any such event, so `ngModel` can't detect the change. This is for example equivalent to setting the element's value property explicitly: `myInput.value = 'foo'`
(Againt, `ngModel` won't be able to detect the change, unless the developer notifies it in some way.)

--
Whether `stepUp()`/`stepDown()` _should_ fire such an event is not 100% clear in the spec, but my interpretation is they shouldn't, as they should qualify as _"changes made to the values of form controls by scripts"_. Relevant spec sections: [dom-input-stepup][1], [common-input-element-events][2]

In any case, regardless if this is a browser bug or expected browser behavior, Angular can't do anything on its own without an event.

--
In your apps, you can work around this behavior by either manually triggering an event or by calling `ngModelController.$setViewValue(...)` (as suggested by @Narretz).

[1]: https://www.w3.org/TR/html5/forms.html#dom-input-stepup
[2]: https://www.w3.org/TR/html5/forms.html#common-input-element-events

**Comment by gkalpak** [comment]
**tl;dr**

In case it wasn't clear, as far as I can understand, `stepUp()`/`stepDown()` are just fancy ways of doing `input.value +/-= 1`, so it makes sense for `ngModel` to behave in the same way.

**Comment by Narretz** [comment]
We can add a note to the docs. We can't really support the methods in Angular easily, so I'd say it won't happen.

**Comment by Narretz** [comment]
Since this is the same as doing `input.value +/-= ` I think adding a note to the docs would possibly be more confusing, as all other inputs have the same behavior.

**Comment by CTCTIPragg** [comment]
I would have to disagree with the conclusion here.

`value += 0.1` is not the same as `input.stepUp()`

The first will give you floating point errors, while the second will not.

https://stackblitz.com/edit/angular-ivy-suyqkx?file=src%2Fapp%2Fapp.component.html

This makes it much more painful to implement custom form controls with increment/decrement functionality.

**Comment by mgol** [comment]
@CTCTIPragg That's an interesting observation. Please note, though, that AngularJS (i.e. version 1.x) is no longer maintained; if there's anything to report for the new Angular, you may want to file an issue at https://github.com/angular/angular/issues.
